### ISSUE TITLE
Feature request: FusedAdamW

### ISSUE THREAD

**Issue by yaysummeriscoming** [NVIDIA/apex#79]
So far I'm really liking apex - no hassle fp16 training.  I've noticed in my experiments that the optimizer does take a not inconsiderable time to execute, so I'm quite interested to try out the new FusedAdam optimizer (once issue 74 is sorted out, that is).  

The thing is, I'm normally using AdamW.  It's small variation of Adam that improves weight decay behaviour.  I understand it's gotten quite popular, for instance fast.ai is using it in all of their work.  Would it also be possible to get a FusedAdamW implementation please?

https://arxiv.org/pdf/1711.05101.pdf
https://www.fast.ai/2018/07/02/adam-weight-decay/

**Comment by mcarilli** [comment]
Are you using Amp?  If so, we're about to include some optimizations targeting Pytorch 1.0 (based on the fact that their C++ API is now stable) which will greatly reduce the cost of checking for Infs and Nans (necessary to implement dynamic loss scaling).

I didn't write the FusedAdam optimizer, but I'll pass along your request.

**Comment by yaysummeriscoming** [comment]
Thanks, currently I'm using FP16_Optimizer instead of Amp.  My usecase is standard imagenet training, using AdamW instead of the SGD.  I understand from the documentation that FP16_Optimizer is best, or should I be using Amp?

Basically I'm after highest throughput.  Lower GPU memory usage is also nice, but throughput is my chief concern.  

I gave dynamic loss scaling a go with FP16_Optimizer, it does seem to be quite a bit slower than using static loss scaling.  I'm glad to hear it'll become faster.  I'm just having a read of the mixed precision training paper again (https://arxiv.org/pdf/1710.03740.pdf), would it otherwise be possible to monitor the mean & standard deviation of the gradients?  Perhaps you could choose a loss scale value such that the distribution stays within acceptable bounds.  Such a check you could also probably just perform once per epoch?

**Comment by FDecaYed** [comment]
Hi @yaysummeriscoming ,
We are looking into adamW as well as some general improvement to fusedAdam. 
Since the fused adam does not support weight decay at all. Do you think we should also support original pytorch adam weight decay?

**Comment by yaysummeriscoming** [comment]
@FDecaYed , that's great news.  Perhaps for completeness it might be a good idea, but in practice I'm always using the AdamW formulation for CNNs.  I've found it to always be superior.

**Comment by EvenOldridge** [comment]
@FDecaYed Any update on this?  I'd also like a fusedAdamW.

**Comment by FDecaYed** [comment]
@yaysummeriscoming @EvenOldridge 
The latest apex.optimizers.FusedAdam already support AdamW. You can use it with optimizers.fp16_optimizer(or amp O2 with manual cast? @mcarilli to confirm)
Sorry for not seeing and replying. I made the change ~ a week after last replied here and forgot to update this thread.

**Comment by yaysummeriscoming** [comment]
Thanks for the update & putting this in!
